ddclient.in

```diff
@@ -2145,6 +2157,157 @@ sub geturl {
     return $reply;
 }
 
+######################################################################
+## geturl -- via curl
+######################################################################
+sub curl_cmd {
+    my @params = @_;
+    my $tmpfile;
+    my %curl_codes = (  ## Subset of error codes from https://curl.haxx.se/docs/manpage.html
+         3 => "URL malformed. The syntax was not correct",
+         5 => "Couldn't resolve proxy. The given proxy host could not be resolved.",
+         6 => "Couldn't resolve host. The given remote host was not resolved.",
+         7 => "Failed to connect to host.",
+        22 => "HTTP page not retrieved. The requested url was not found or returned another error.",
+        28 => "Operation timeout. The specified time-out period was reached according to the conditions.",
+        35 => "SSL connect error. The SSL handshaking failed.",
+        47 => "Too many redirects. When following redirects, curl hit the maximum amount.",
+        51 => "The peer's SSL certificate or SSH MD5 fingerprint was not OK.",
+        58 => "Problem with the local certificate.",
+        60 => "Peer certificate cannot be authenticated with known CA certificates.",
+        67 => "The user name, password, or similar was not accepted and curl failed to log in.",
+        77 => "Problem with reading the SSL CA cert (path? access rights?).",
+        78 => "The resource referenced in the URL does not exist.",
+    );
+
+    return '' if (scalar(@params) == 0); ## no parameters provided
+
+    # debug("CURL:\r\n%s\r\n",$params); ## Uncomment only if required as may expose credentials
+
+    eval "no warnings 'all'; use File::Temp;";
+    if (!$@) {
+        ## Use system File::Temp to generate filename...
+        $tmpfile = File::Temp->new(TMPDIR => 1,  ## in system temp dir
+                                   UNLINK => 0,  ## we will delete it ourselves
+                                 TEMPLATE => 'ddclient_XXXXXXXXXX');
+        ## File::Temp->new() automatically opens the file for us.
+    } else {
+        ## System does not have File::Temp so generate ourselves...
+        eval "no warnings 'all'; use File::Spec;";
+        $tmpfile = ((!$@) ? File::Spec->tmpdir() : '/tmp').'/ddclient_';
+        my @chars = ("A".."Z", "a".."z", "0".."9");
+        $tmpfile .= $chars[rand @chars] for 1..10;
+        open($tmpfile, '>', $tmpfile) or fatal("Falied to open $tmpfile for curl");
+    }
+    chmod(0600,$tmpfile);
+    debug("CURL Tempfile: %s",$tmpfile);
+    local $\ = "\n";  ## Terminate the file,
+    local $, = "\n";  ## and each parameter, with a newline.
+    print($tmpfile @params);
+    close($tmpfile);
+
+    ## Make sure to delete temp file on Ctrl-C or SIGTERM
+    local $SIG{'INT'} = sub { unlink($tmpfile); exit 2 };
+    local $SIG{'TERM'} = sub { unlink($tmpfile); exit 15 };
+    my $reply = qx{ /usr/bin/curl --config '$tmpfile' 2>/dev/null; };
```

Comment:
What does `curl` output? Is it the full HTTP response including the status line and headers? The return value should match what `fetch_via_socket_io` returns. We can change `fetch_via_socket_io` if that is easier than changing `fetch_via_curl`.

Reply:
I'm still working on #170. It's tedious to audit every function to make sure I'm not breaking anything, so it'll be a while.

Let's leave this unresolved until either I finish #170 or this is changed to return the full response.

Reply:
The vast majority of places where it is used it just checks `!header_ok()` and move on if it comes back not ok... which means anything other than an http 200.  Those are easy to handle and in almost every case the header_ok function call would not even be executed (because geturl would have returned '' and that is checked before header).

In the few places that get a positive reply the first thing they do is strip off the header, so it is thrown away.

My point is, it's not as hard as you think.

Reply:
Curl version now returns HTTP headers for compatibility.  We can address in #170 which you can assign to me if you want after this is merged.